src/lib.rs

```diff
@@ -0,0 +1,107 @@
+use std::env;
+
+use log::info;
+use tokio::sync::mpsc;
+
+use trin_core::cli::TrinConfig;
+use trin_core::jsonrpc::launch_jsonrpc_server;
+use trin_core::portalnet::protocol::{
+    HistoryNetworkEndpoint, JsonRpcHandler, PortalEndpoint, PortalnetConfig, PortalnetProtocol,
+    StateNetworkEndpoint,
+};
+use trin_history::HistoryRequestHandler;
+use trin_state::StateRequestHandler;
+
+pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    println!("Launching trin");
+
+    let trin_config = TrinConfig::new();
+    trin_config.display_config();
+
+    let infura_project_id = match env::var("TRIN_INFURA_PROJECT_ID") {
+        Ok(val) => val,
+        Err(_) => panic!(
+            "Must supply Infura key as environment variable, like:\n\
+            TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
+        ),
+    };
+
+    let bootnode_enrs = trin_config
+        .bootnodes
+        .iter()
+        .map(|nodestr| nodestr.parse().unwrap())
+        .collect();
+
+    let portalnet_config = PortalnetConfig {
+        external_addr: trin_config.external_addr,
+        private_key: trin_config.private_key.clone(),
+        listen_port: trin_config.discovery_port,
+        bootnode_enrs,
+        ..Default::default()
+    };
+
+    let mut state_tx = None;
+    let mut history_tx = None;
+    let mut state_handler: Option<StateRequestHandler> = None;
+    let mut history_handler: Option<HistoryRequestHandler> = None;
+
+    if trin_config.networks.iter().any(|val| val == "history") {
+        let (raw_history_tx, history_rx) = mpsc::unbounded_channel::<HistoryNetworkEndpoint>();
+        history_tx = Some(raw_history_tx);
+        history_handler = match trin_history::initialize(history_rx) {
+            Ok(val) => Some(val),
+            Err(msg) => panic!("Error while initializing history network: {:?}", msg),
+        }
+    }
+
+    if trin_config.networks.iter().any(|val| val == "state") {
+        let (raw_state_tx, state_rx) = mpsc::unbounded_channel::<StateNetworkEndpoint>();
+        state_tx = Some(raw_state_tx);
+        state_handler = match trin_state::initialize(state_rx) {
+            Ok(val) => Some(val),
+            Err(msg) => panic!("Error while initializing state network: {:?}", msg),
+        }
+    }
```

Comment:
Perhaps the following is a cleaner way than to declare the `mut` variables above?

```suggestion
    let (state_tx, state_handler) = if trin_config.networks.iter().any(|val| val == "state") {
        let (state_tx, state_rx) = mpsc::unbounded_channel::<StateNetworkEndpoint>();
        state_handler = match trin_state::initialize(state_rx) {
            Ok(val) => val,
            Err(msg) => panic!("Error while initializing state network: {:?}", msg),
        }
        (Some(state_tx), Some(state_handler))
    } else { None, None };
```

Similarly for history.

Reply:
I think I liked this one even better:

```rust
for network in trin_config.networks.iter() {
        match network {....
```

Reply:
It's very possible I'm missing how to accomplish this, but I wasn't able to find a way to use the `iter()` strategy @ogenev suggested without instantiating `None` mut variables in advance. Imo those `mut` variables are grosser than the clunky iteration handling - but I have no strong feelings towards either approach.